File: src/main/java/view/utils/GridBagPanelAdder.java
package view.utils;

import javax.swing.*;
import java.awt.*;

public class GridBagPanelAdder {

	private int gridX = 0, gridY = 0, wideX = 1, wideY = 1,
			north = 0, east = 0, south = 0, west = 0;
	private double weightX = 0, weightY = 0;
	private int fill = GridBagConstraints.BOTH;

	public GridBagPanelAdder setFill ( int fill ) {
		this.fill = fill;
		return this;
	}

	public GridBagPanelAdder setPadding( int north, int east, int south, int west ) {
		this.north = north;
		this.west = west;
		this.south = south;
		this.east = east;
		return this;
	}

	public GridBagPanelAdder setPosition( int gridX, int gridY ) {
		this.gridX = gridX;
		this.gridY = gridY;
		return this;
	}

	public GridBagPanelAdder setWeight( double weightX, double weightY ) {
		this.weightX = weightX;
		this.weightY = weightY;
		return this;
	}

	public GridBagPanelAdder setWideness( int wideX, int wideY ) {
		this.wideX = wideX;
		this.wideY = wideY;
		return this;
	}

	public void addToPanel( JPanel panel, Component item ) {
		GridBagConstraints constraints = new GridBagConstraints( );
		constraints.anchor = GridBagConstraints.NORTHEAST;
		constraints.fill = fill;
		constraints.weightx = weightX;
		constraints.weighty = weightY;
		constraints.gridx = gridX;
		constraints.gridy = gridY;
		constraints.gridwidth = wideX;
		constraints.gridheight = wideY;
		constraints.insets = new Insets( north, west, south, east );
		panel.add( item, constraints );
	}
}

File: client3/src/main/kotlin/view/util/GridBagPanelAdder.kt
package view.util

import java.awt.Component
import java.awt.Container
import java.awt.GridBagConstraints
import java.awt.GridBagConstraints.BOTH
import java.awt.GridBagConstraints.NORTHEAST
import java.awt.Insets

/**
 * An utility builder to comfortably locate component using GridBagLayout
 *
 * @author Paolo Baldini
 */
class GridBagPanelAdder {
    private var gridX = 0
    private var gridY = 0
    private var wideX = 1
    private var wideY = 1
    private var north = 0
    private var east = 0
    private var south = 0
    private var west = 0
    private var weightX = 0.0
    private var weightY = 0.0
    private var fill = BOTH

    fun xPos(value: Int) = apply { gridX = value }
    fun yPos(value: Int) = apply { gridY = value }
    fun position(gridX: Int, gridY: Int) = apply { this.gridX = gridX; this.gridY = gridY }

    fun xWeight(value: Double) = apply { weightX = value }
    fun yWeight(value: Double) = apply { weightY = value }
    fun weight(weightX: Double, weightY: Double) = apply { this.weightX = weightX; this.weightY = weightY }

    fun xWide(value: Int) = apply { wideX = value }
    fun yWide(value: Int) = apply { wideY = value }
    fun wideness(wideX: Int, wideY: Int) = apply { this.wideX = wideX; this.wideY = wideY }

    fun north(value: Int) = apply { north = value }
    fun east(value: Int) = apply { east = value }
    fun south(value: Int) = apply { south = value }
    fun west(value: Int) = apply { west = value }
    fun padding(north: Int, east: Int, south: Int, west: Int): GridBagPanelAdder {
        this.north = north
        this.west = west
        this.south = south
        this.east = east
        return this
    }

    fun fill(value: Int) = apply { fill = value }

    fun <T: Component>addTo(panel: Container, item: T): T {
        val constraints = GridBagConstraints().apply { anchor = NORTHEAST }
            .apply { fill = this@GridBagPanelAdder.fill }
            .apply { weightx = weightX; weighty = weightY }
            .apply { gridx = gridX; gridy = gridY }
            .apply { gridwidth = wideX; gridheight = wideY }
            .also { it.insets = Insets(north, west, south, east) }
        panel.add(item, constraints)
        return item
    }
}